Observer pattern with weighted random selection

AyeshaSiddique01 opened the pull request.

AyeshaSiddique01 committed:
build: observer pattern with weighted random selection

AyeshaSiddique01 committed:
refactor

AyeshaSiddique01 committed:
refactor: topicstore removed

AyeshaSiddique01 committed:
refactor: topics added in student class

faisal05sa commented on observer_weighted_random_Selection.py:
```diff
@@ -0,0 +1,114 @@
+        observer_index = self.observers.index(observer)
+        self.observers.remove(observer)
+        observer.observable = None
+        self.weights.remove(self.weights[observer_index])
```
You should directly remove the element from that index. 
```python
self.weights.pop(observer_index)
```

AyeshaSiddique01 committed:
refactor: topic is not being removed by student
sets student's weight to 0
Print the selected topics and see if all students have uniquely selected the topics

AyeshaSiddique01 committed:
refactor: interfaces imported from observer.py

faisal05sa reviewed (changes requested):
I think you can inherit the Observer and Observable classes and extend the functionality wherever required.

AyeshaSiddique01 committed:
refactor: code refactored

AyeshaSiddique01 committed:
refactor: extended functionality from observer.py
